pkg/file/splitter/internal/job.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"hash"
+	"os"
+
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bmt"
+	bmtlegacy "github.com/ethersphere/bmt/legacy"
+	"golang.org/x/crypto/sha3"
+)
+
+// hashFunc is a hasher factory used by the bmt hasher
+func hashFunc() hash.Hash {
+	return sha3.NewLegacyKeccak256()
+}
+
+// SimpleSplitterJob encapsulated a single splitter operation, accepting blockwise
+// writes of data whose length is defined in advance.
+//
+// After the job is constructed, Write must be called with up to ChunkSize byte slices
+// until the full data length has been written. The Sum should be called which will
+// return the SwarmHash of the data.
+//
+// Called Sum before the last Write, or Write after Sum has been called, may result in
+// error and will may result in undefined result.
+type SimpleSplitterJob struct {
+	ctx        context.Context
+	store      storage.Storer
+	spanLength int64    // target length of data
+	length     int64    // number of bytes written to the data level of the hasher
+	sumCounts  []int    // number of sums performed, indexed per level
+	cursors    []int    // section write position, indexed per level
+	hasher     bmt.Hash // underlying hasher used for hashing the tree
+	buffer     []byte   // keeps data and hashes, indexed by cursors
+	logger     logging.Logger
+}
+
+// NewSimpleSplitterJob creates a new SimpleSplitterJob.
+//
+// The spanLength is the length of the data that will be written.
+func NewSimpleSplitterJob(ctx context.Context, store storage.Storer, spanLength int64) *SimpleSplitterJob {
+
+	p := bmtlegacy.NewTreePool(hashFunc, swarm.Branches, bmtlegacy.PoolSize)
+	j := &SimpleSplitterJob{
+		ctx:        ctx,
+		store:      store,
+		spanLength: spanLength,
+		sumCounts:  make([]int, 9),
+		cursors:    make([]int, 9),
+		hasher:     bmtlegacy.New(p),
+		buffer:     make([]byte, swarm.ChunkSize*9),
+		logger:     logging.New(os.Stderr, 6),
+	}
+	return j
+}
+
+// Write adds data to the file splitter.
+func (j *SimpleSplitterJob) Write(b []byte) (int, error) {
+	if len(b) > swarm.ChunkSize {
+		return 0, fmt.Errorf("Write must be called with a maximum of %d bytes", swarm.ChunkSize)
```

Comment:
i'd consider even being more harsh here at this point and even `panic`, so we have an immediate idea that some grave mistake such as this has occured. @janos WDYT?

Reply:
Since this is an exported Method, I would go with just an error.